Convert format script to Dart

gspencergoog opened the pull request:
## Description

This converts the ci/format.sh script to Dart script that uses `process_runner` to multi-process the clang-format, diffs, and grepping needed to do the formatting changes.

It also will (by default) only check the formatting of changed files.

The user can optionally check all files (`--all-files`) or do only some types of checks with `--check`. `--verbose` prints the versions of the tools used for Clang format and Java format.

Specifying `--fix` will cause any formatting errors that would have been detected to be fixed.

## Related Issues

 - Fixes https://github.com/flutter/flutter/issues/55472

stuartmorgan commented:
This will fix/obsolete https://github.com/flutter/flutter/issues/55472

stuartmorgan commented:
Could you add `.bat` versions of the wrappers as well? That would allow running the formatter on Windows, which would be nice.

gspencergoog commented:
Sure, I'll .bat versions too.

gspencergoog commented:
@stuartmorgan Actually, that's not possible yet: the Windows build still relies on MSVC, not clang, so clang-format isn't available to run, nor is the Google Java formatter. I can check for trailing whitespace, and fix GN files, and that's about it.

gspencergoog commented:
Actually, I can remove whitespace.  GN doesn't seem to support formatting on Windows either.

dnfield commented on ci/bin/format.dart:
```diff
@@ -0,0 +1,795 @@
+  FormattingException(this.message, [this.result]);
+
+  final String message;
+  final ProcessResult /*?*/ result;
```
Aren't we on an NNBD enabled SDK already?

gspencergoog replied:
Yes, but when I tried to turn that on, it wanted all of the dependent packages to be NNBD too. Maybe I was holding it wrong, though.

dnfield replied:
Ahh. Maybe we can just migrate all the stuff  in here at once then.  That's fine.

dnfield commented on ci/bin/format.dart:
```diff
@@ -0,0 +1,795 @@
+
+  @override
+  String toString() {
+    String output = runtimeType.toString();
```
nit: consider using a StringBuffer.

dnfield commented on ci/bin/format.dart:
```diff
@@ -0,0 +1,795 @@
+    String output = runtimeType.toString();
+    output += ': $message';
+    final String stderr = result?.stderr as String;
+    if (stderr.isNotEmpty) {
```
stderr might be null right?

dnfield commented on ci/bin/format.dart:
```diff
@@ -0,0 +1,795 @@
+          "Unknown operating system: don't know how to run clang-format here.");
+    }
+    clangFormat = File(
+        path.join(srcDir.absolute.path, 'buildtools', clangOs, 'clang', 'bin', 'clang-format'));
```
nit: add a trailing comma for nicer formatting

gspencergoog replied:
Done.

dnfield commented on ci/bin/format.dart:
```diff
@@ -0,0 +1,795 @@
+    );
+    final List<WorkerJob> jobs = patches.map<WorkerJob>((String patch) {
+      return WorkerJob(<String>['patch', '-p0'],
+          stdinRaw: codeUnitsAsStream(patch.codeUnits), failOk: true);
```
nit: trailing comma

gspencergoog replied:
Done.

stuartmorgan commented:
> the Windows build still relies on MSVC, not clang

Well, it relies on both, since we're using clang-cl. But it does look like clang-format isn't in our package. Maybe once we're using Fuchsia's we can get it from there. (Although it seems like it should be in the Chromium bundle, so maybe I'm just missing it?)

> GN doesn't seem to support formatting on Windows either.

Really? I could have sworn I've formatted with it. Maybe I'm using a different copy than the bundled one.

dnfield commented on ci/bin/format.dart:
```diff
@@ -0,0 +1,795 @@
+      clangJobs.add(WorkerJob(<String>[clangFormat.path, '--style=file', file.trim()]));
+    }
+    final ProcessPool clangPool =
+        ProcessPool(processRunner: _processRunner, printReport: namedReport('clang-format'));
```
trailing comma

gspencergoog replied:
Done.

dnfield commented on ci/bin/format.dart:
```diff
@@ -0,0 +1,795 @@
+    await for (final WorkerJob completedJob in completedClangFormats) {
+      if (completedJob.result != null && completedJob.result.exitCode == 0) {
+        diffJobs.add(WorkerJob(<String>['diff', '-u', completedJob.command.last, '-'],
+            stdinRaw: codeUnitsAsStream(completedJob.result.stdoutRaw), failOk: true));
```
trailing commas would help here

dnfield commented on ci/bin/format.dart:
```diff
@@ -0,0 +1,795 @@
+      }
+    }
+    final ProcessPool diffPool =
+        ProcessPool(processRunner: _processRunner, printReport: namedReport('diff'));
```
trailing commas

gspencergoog replied:
Done.

dnfield commented on ci/bin/format.dart:
```diff
@@ -0,0 +1,795 @@
+      formatJobs.add(WorkerJob(<String>['java', '-jar', googleJavaFormatJar.path, file.trim()]));
+    }
+    final ProcessPool formatPool =
+        ProcessPool(processRunner: _processRunner, printReport: namedReport('Java format'));
```
,

gspencergoog replied:
Done.

dnfield commented on ci/bin/format.dart:
```diff
@@ -0,0 +1,795 @@
+    await for (final WorkerJob completedJob in completedClangFormats) {
+      if (completedJob.result != null && completedJob.result.exitCode == 0) {
+        diffJobs.add(WorkerJob(<String>['diff', '-u', completedJob.command.last, '-'],
+            stdinRaw: codeUnitsAsStream(completedJob.result.stdoutRaw), failOk: true));
```
,

gspencergoog replied:
Done.

dnfield commented on ci/bin/format.dart:
```diff
@@ -0,0 +1,795 @@
+
+  Future<int> _runGnCheck({@required bool fixing}) async {
+    final List<String> cmd = <String>[
+      'python',
```
doesn't have to be this patch, but it's a little strange to me to have this utility exist in python when everything else in here is Dart or some other native thing maintained by others. I suspect we really want this to use the same file inding logic in here rather than being duplicated.

The call to `gn format` just needs a `-w` for fixing. Up to you whether you want to include that in this patch or just keep it as a rewrite o the bash script for now, but we should probably at least add a todo for it.

dnfield commented on ci/bin/format.dart:
```diff
@@ -0,0 +1,795 @@
+  }
+
+  Future<List<String>> _getWhitespaceFailures() async {
+    final List<String> files = await getFileList(<String>['*.dart']);
```
Why aren't we doing this on _all_ source files?

dnfield commented on ci/bin/format.dart:
```diff
@@ -0,0 +1,795 @@
+  }
+}
+
+Future<String> _runGit(List<String> args, ProcessRunner processRunner) async {
+  final ProcessRunnerResult result = await processRunner.runProcess(<String>['git', ...args]);
+  return result.stdout;
+}
```
Can we somehow merge this an `runGit` above?

gspencergoog commented:
@stuartmorgan Actually, I was holding it wrong: the "binary" it wants is the one in `third_party\gn\gn`, not `tools\gn.bat`.  Looks like it might work.

I'm still not sure it's worth the effort to make it run on Windows if we can only do GN files and trailing whitespace. What do you think?

stuartmorgan commented:
> @stuartmorgan Actually, I was holding it wrong: the "binary" it wants is the one in `third_party\gn\gn`, not `tools\gn.bat`. Looks like it might work.
> 
> I'm still not sure it's worth the effort to make it run on Windows if we can only do GN files and trailing whitespace. What do you think?

I think minimal support now and adding clang-format later (there's no reason we can't support it, we just need to make sure we get clang-format when we get other Windows clang binaries) would be better than nothing, but we could also just do a full pass of Windows support later.

gspencergoog commented:
@stuartmorgan I added support for checking/fixing GN files and whitespace on Windows, and added a `format.bat` file.  When we get clang support, we can easily turn it on by changing the `formatCheckNames` function to allow more check types on Windows.

stuartmorgan commented:
Awesome, thanks for rolling that into this work :)
